test/integration/test_vault_file_source.py

```diff
@@ -0,0 +1,90 @@
+import os
+import tempfile
+
+from galaxy.security.vault import UserVaultWrapper
+from galaxy_test.base import api_asserts
+from galaxy_test.driver import integration_util
+
+
+SCRIPT_DIRECTORY = os.path.abspath(os.path.dirname(__file__))
+FILE_SOURCES_VAULT_CONF = os.path.join(SCRIPT_DIRECTORY, "file_sources_conf_vault.yml")
+VAULT_CONF = os.path.join(SCRIPT_DIRECTORY, "vault_conf.yml")
+
+
+class VaultFileSourceIntegrationTestCase(integration_util.IntegrationTestCase):
+    USER_1_APP_VAULT_ENTRY = "randomvaultuser1@universe.com"
```

Comment:
This integration test is really good - it would be even better if you actually performed an upload I think. I think it is the case that the remote job for instance won't attempt to read the database because the config will be stored with the secrets expanded - but I think if the test verified that it would be great.

I also wish you weren't digging into app and creating a UserVaultWrapper - if there were API endpoints for that this would be better. But I guess there aren't - or can the UserVaultWrapper thing be set through the preferences API in some way?

Reply:
@jmchilton That scenario is already being tested, because the test makes sure the secret is in fact expanded.

The only reason to dig directly into the app is to set up the text fixtures - I couldn't think of another way to seed the necessary vault data, since the test needs to ensure that the correct underlying storage is used.

Thanks for reviewing this.

Reply:
I am probably missing this, but I don't see where the test is running a job, and so I can't quite tell how/if decrypted secrets are passed to jobs. When we run a job we serialize the file sources for a user ... like John I would also assume this just works, but it would be nice to have a test that does that.

Reply:
I did dig into this, and indeed I think this works correctly. I tried it with this test:
```diff
diff --git a/test/integration/test_vault_file_source.py b/test/integration/test_vault_file_source.py
index 4b0ebcbc32..1f7c73b0d2 100644
--- a/test/integration/test_vault_file_source.py
+++ b/test/integration/test_vault_file_source.py
@@ -1,8 +1,10 @@
+import json
 import os
 import tempfile
 
 from galaxy.security.vault import UserVaultWrapper
 from galaxy_test.base import api_asserts
+from galaxy_test.base.populators import DatasetPopulator
 from galaxy_test.driver import integration_util
 
 
@@ -15,6 +17,10 @@ class VaultFileSourceIntegrationTestCase(integration_util.IntegrationTestCase):
     USER_1_APP_VAULT_ENTRY = "randomvaultuser1@universe.com"
     USER_2_APP_VAULT_ENTRY = "randomvaultuser2@universe.com"
 
+    def setUp(self):
+        super().setUp()
+        self.dataset_populator = DatasetPopulator(self.galaxy_interactor)
+
     @classmethod
     def handle_galaxy_config_kwds(cls, config):
         config["file_sources_config_file"] = FILE_SOURCES_VAULT_CONF
@@ -52,6 +58,36 @@ class VaultFileSourceIntegrationTestCase(integration_util.IntegrationTestCase):
             list_response = self.galaxy_interactor.get("remote_files", data)
             api_asserts.assert_status_code_is(list_response, 404)
 
+    def test_upload_file_from_remote_source(self):
+        with self._different_user(email=self.USER_1_APP_VAULT_ENTRY):
+            app = self._app
+            user = app.model.context.query(app.model.User).filter(
+                app.model.User.email == self.USER_1_APP_VAULT_ENTRY).first()
+            user_vault = UserVaultWrapper(self._app.vault, user)
+            # use a valid symlink path so the posix list succeeds
+            user_vault.write_secret('posix/root_path', app.config.user_library_import_symlink_allowlist[0])
+            data = {"target": "gxfiles://test_user_vault"}
+            with open(os.path.join(app.config.user_library_import_symlink_allowlist[0], 'a_file'), 'w') as fh:
+                fh.write('I require access to the vault')
+            list_response = self.galaxy_interactor.get("remote_files", data)
+            api_asserts.assert_status_code_is_ok(list_response)
+            remote_files = list_response.json()
+            assert len(remote_files) == 1
+            with self.dataset_populator.test_history() as history_id:
+                element = dict(src="url", url="gxfiles://test_user_vault/a_file")
+                target = {
+                    "destination": {"type": "hdas"},
+                    "elements": [element],
+                }
+                targets = json.dumps([target])
+                payload = {
+                    "history_id": history_id,
+                    "targets": targets,
+                }
+                new_dataset = self.dataset_populator.fetch(payload, assert_ok=True).json()["outputs"][0]
+                content = self.dataset_populator.get_history_dataset_content(history_id, dataset=new_dataset)
+                assert content == "I require access to the vault", content
+
     def test_vault_secret_per_app_in_file_source(self):
         """
         This file source performs an app level vault lookup. Although the secret stored for the first user is a

```
Couldn't you do the test setup in the the same way you do it in `test_extra_prefs_vault_storage`, at least for the user vault test ?